llama-index-packs/llama-index-packs-self-discover/llama_index/packs/self_discover/base.py

```diff
@@ -49,96 +58,111 @@
+)
 
-    def add_modules(self) -> None:
-        """Adds necessary modules and their configurations to the pipeline."""
-        self.pipeline.add_modules(
-            {
-                "input": InputComponent(),
-                "select_llm": self.llm,
-                "adapt_llm": self.llm,
-                "implement_llm": self.llm,
-                "reasoning_llm": self.llm,
-                "select_prompt_template": self.select_prompt_template,
-                "adapt_prompt_template": self.adapt_prompt_template,
-                "implement_prompt_template": self.implement_prompt_template,
-                "reasoning_prompt_template": self.reasoning_prompt_template,
-            }
-        )
 
-    def setup_links(self) -> None:
-        """Defines the connections (links) between the different pipeline modules."""
-        # STAGE-1: SELECT subset of reasoning Modules.
-        self.pipeline.add_link(
-            "input", "select_prompt_template", src_key="task", dest_key="task"
-        )
-        self.pipeline.add_link(
-            "input",
-            "select_prompt_template",
-            src_key="reasoning_modules",
-            dest_key="reasoning_modules",
-        )
-        self.pipeline.add_link("select_prompt_template", "select_llm")
+class GetModulesEvent(Event):
+    """Event to get modules."""
 
-        # STAGE-1: ADAPT selected reasoning modules to the task.
-        self.pipeline.add_link(
-            "select_llm", "adapt_prompt_template", dest_key="selected_modules"
-        )
-        self.pipeline.add_link(
-            "input", "adapt_prompt_template", src_key="task", dest_key="task"
-        )
-        self.pipeline.add_link("adapt_prompt_template", "adapt_llm")
+    task: str
+    modules: str
 
-        # STAGE-1: IMPLEMENT provides reasoning structure for the task.
-        self.pipeline.add_link(
-            "adapt_llm", "implement_prompt_template", dest_key="adapted_modules"
-        )
-        self.pipeline.add_link(
-            "input", "implement_prompt_template", src_key="task", dest_key="task"
-        )
-        self.pipeline.add_link("implement_prompt_template", "implement_llm")
 
-        # STAGE-2: Uses the generated reasoning structure for the task to generate an answer.
-        self.pipeline.add_link(
-            "implement_llm", "reasoning_prompt_template", dest_key="reasoning_structure"
+class RefineModulesEvent(Event):
+    """Event to refine modules."""
+
+    task: str
+    refined_modules: str
+
+
+class ReasoningStructureEvent(Event):
+    """Event to create reasoning structure."""
+
+    task: str
+    reasoning_structure: str
+
+
+class SelfDiscoverWorkflow(Workflow):
+    """Self discover workflow."""
+
+    @step(pass_context=True)
+    async def get_modules(self, ctx: Context, ev: StartEvent) -> GetModulesEvent:
+        """Get modules step."""
+        # get input data, store llm into ctx
+        task = ev.get("task")
+        llm: LLM = ev.get("llm")
```

Comment:
For workflows with a single entry point, might be more friendly to confirm these are actually properly passed in, and raise an appropriate error if not